skylog/storage/gcp/pubsub/main.go

```diff
@@ -0,0 +1,83 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package contains a binary that sends Merkle tree building jobs to a
+// collection of workers through a GCP Pub/Sub topic.
+package main
+
+import (
+	"context"
+	"flag"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+	pb "github.com/google/trillian/skylog/storage/gcp/gcppb"
+	"golang.org/x/time/rate"
+)
+
+var (
+	projectID  = flag.String("project", "skylog-test", "The GCP project ID")
+	psTopic    = flag.String("topic", "build-jobs", "The Pub/Sub topic for build jobs")
+	treeID     = flag.Int64("tree_id", 1, "The ID of the tree under construction")
+	beginIndex = flag.Int64("begin", 0, "The beginning of the tree building range")
+	endIndex   = flag.Int64("end", 0, "The ending of the tree building range")
+	maxJobSize = flag.Int("job_size", 256, "The maximal number of entries in a build job")
+	maxRate    = flag.Float64("rate", 100000, "The average rate of adding entries per second")
+)
+
+func runSender(ctx context.Context, cli *pubsub.Client) error {
+	top := cli.Topic(*psTopic)
+	defer top.Stop()
+
+	var results []*pubsub.PublishResult
+
+	lim := rate.NewLimiter(rate.Limit(*maxRate), *maxJobSize)
+	for index, next := *beginIndex, int64(0); index < *endIndex; index = next {
+		if next = index + int64(*maxJobSize); next > *endIndex {
```

Comment:
I'd break this out into two lines for clarity - there's a "leaking" side-effect in the if (`next = ...` as opposed to `next := ...`) which is unusual.